opt: incorporate operator volatility

RaduBerinde opened the pull request:
Incorporate volatility of operators (unary, binary, comparison) into the
VolatilitySet property.

Unfortunately, this modifies most plans as pretty much everything is
"immutable". Perhaps once this work is behind us we will want to hide this
information from plans in most cases.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49923)
<!-- Reviewable:end -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 75 of 75 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49923#-:-M9J0dsv1iobtrfRK-lm:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner and @RaduBerinde)

___
*[pkg/sql/opt/memo/logical_props_builder.go, line 1376 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49923#-M9Iyyqv-3VP9yDm13y5:-M9Iyyqv-3VP9yDm13y6:bexam63) ([raw file](https://github.com/cockroachdb/cockroach/blob/3968765fffe3bd4a357bc8551d69cc6ab1aa7e0e/pkg/sql/opt/memo/logical_props_builder.go#L1376)):*
> ```Go
> 		// non-zero constant.
> 		//
> 		// TODO(radu): clean up this special case.
> ```

Maybe expand on this TODO a bit to note that this case should eventually be removed so all division expressions will have the same logic as the `IsBinaryOp` logic below.
___
*[pkg/sql/opt/memo/logical_props_builder.go, line 1422 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49923#-M9Iy3FmErZRLsTc77K5:-M9Iy3FnBQogPpWprzyz:b-495g5z) ([raw file](https://github.com/cockroachdb/cockroach/blob/3968765fffe3bd4a357bc8551d69cc6ab1aa7e0e/pkg/sql/opt/memo/logical_props_builder.go#L1422)):*
> ```Go
> 			o, ok := FindUnaryOverload(e.Op(), inputType)
> 			if !ok {
> 				panic(errors.AssertionFailedf("unary overload found (%s, %s)", e.Op(), inputType))
> ```

found -> not found


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49923)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @RaduBerinde)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/49923#-M9Ji63i0fXPdXQu9lvz:-M9Ji63i0fXPdXQu9lw-:bg7igkv) (no related file):*
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
opt: incorporate operator volatility

Incorporate volatility of operators (unary, binary, comparison) into the
VolatilitySet property.

Unfortunately, this modifies most plans as pretty much everything is
"immutable". Perhaps once this work is behind us we will want to hide this
information from plans in most cases.

Release note: None

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49923)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @RaduBerinde and @rytaft)

___
*[pkg/sql/opt/memo/logical_props_builder.go, line 1376 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49923#-M9Iyyqv-3VP9yDm13y5:-M9JyelG30qg7lFSO03E:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3968765fffe3bd4a357bc8551d69cc6ab1aa7e0e/pkg/sql/opt/memo/logical_props_builder.go#L1376)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Maybe expand on this TODO a bit to note that this case should eventually be removed so all division expressions will have the same logic as the `IsBinaryOp` logic below.
</blockquote></details>

Done.
___
*[pkg/sql/opt/memo/logical_props_builder.go, line 1422 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/49923#-M9Iy3FmErZRLsTc77K5:-M9JykWM8nrjXifFQ0vC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3968765fffe3bd4a357bc8551d69cc6ab1aa7e0e/pkg/sql/opt/memo/logical_props_builder.go#L1422)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

found -> not found
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 9 of 9 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49923)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @RaduBerinde)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
